File: grasshopper_mcp_server/src/grasshoppermcp/tools/get_document_info.py
"""Get information about the active Grasshopper document."""

from mcp.server.fastmcp import Context
from grasshoppermcp.server import get_grasshopper_connection, mcp, logger
from typing import Dict, Any


@mcp.tool()
def get_gh_document_info(ctx: Context) -> Dict[str, Any]:
    """
    Get information about the active Grasshopper document.

    Returns:
    - has_document: Whether a GH document is open
    - file_path: Path to the GH file (or "(unsaved)")
    - object_count: Total number of objects on canvas
    - component_count: Number of components
    - parameter_count: Number of standalone parameters
    - group_count: Number of groups
    - components_by_category: Breakdown of components by category

    Example:
        info = get_gh_document_info()
        if info["has_document"]:
            print(f"Document has {info['component_count']} components")
    """
    try:
        gh = get_grasshopper_connection()
        return gh.send_command("get_gh_document_info", {})
    except Exception as e:
        logger.error(f"Error getting GH document info: {str(e)}")
        return {"success": False, "message": str(e)}

File: grasshopper_mcp_plugin/Functions/GetDocumentInfo.cs
using System;
using System.Linq;
using Grasshopper;
using Grasshopper.Kernel;
using Newtonsoft.Json.Linq;

namespace GrasshopperMCPPlugin.Functions;

public partial class GrasshopperMCPFunctions
{
    /// <summary>
    /// Get information about the active Grasshopper document.
    /// </summary>
    public JObject GetDocumentInfo(JObject parameters)
    {
        var doc = Instances.ActiveCanvas?.Document;

        if (doc == null)
        {
            return new JObject
            {
                ["has_document"] = false,
                ["message"] = "No active Grasshopper document"
            };
        }

        // Count components by category
        var componentsByCategory = doc.Objects
            .OfType<IGH_Component>()
            .GroupBy(c => c.Category ?? "Unknown")
            .ToDictionary(g => g.Key, g => g.Count());

        // Count parameters
        var parameterCount = doc.Objects.OfType<IGH_Param>().Count();

        // Get document properties
        var result = new JObject
        {
            ["has_document"] = true,
            ["file_path"] = doc.FilePath ?? "(unsaved)",
            ["is_modified"] = doc.IsModified,
            ["object_count"] = doc.ObjectCount,
            ["component_count"] = doc.Objects.OfType<IGH_Component>().Count(),
            ["parameter_count"] = parameterCount,
            ["group_count"] = doc.Objects.OfType<Grasshopper.Kernel.Special.GH_Group>().Count(),
            ["components_by_category"] = JObject.FromObject(componentsByCategory)
        };

        return result;
    }
}
